bindings/matrix-sdk-crypto-js/src/backup_recovery_key.rs

```diff
@@ -0,0 +1,176 @@
+//! Megolm backup types
+
+use std::{collections::HashMap, iter, ops::DerefMut};
+
+use hmac::Hmac;
+use js_sys::{JsString, JSON};
+use matrix_sdk_crypto::{backups::MegolmV1BackupKey as InnerMegolmV1BackupKey, store::RecoveryKey};
+use pbkdf2::pbkdf2;
+use rand::{distributions::Alphanumeric, thread_rng, Rng};
+use serde_wasm_bindgen;
+use sha2::Sha512;
+use wasm_bindgen::prelude::*;
+use zeroize::Zeroize;
+
+/// The private part of the backup key, the one used for recovery.
+#[derive(Debug)]
+#[wasm_bindgen]
+pub struct BackupRecoveryKey {
+    pub(crate) inner: RecoveryKey,
+    pub(crate) passphrase_info: Option<PassphraseInfo>,
+}
+
+/// Struct containing info about the way the backup key got derived from a
+/// passphrase.
+#[derive(Debug, Clone)]
+#[wasm_bindgen]
+pub struct PassphraseInfo {
+    /// The salt that was used during key derivation.
+    #[wasm_bindgen(getter_with_clone)]
+    pub private_key_salt: JsString,
+    /// The number of PBKDF rounds that were used for key derivation.
+    pub private_key_iterations: i32,
+}
+
+/// The public part of the backup key.
+#[derive(Debug, Clone)]
+#[wasm_bindgen]
+pub struct MegolmV1BackupKey {
+    inner: InnerMegolmV1BackupKey,
+    passphrase_info: Option<PassphraseInfo>,
+}
+
+#[wasm_bindgen]
+impl MegolmV1BackupKey {
+    /// The actual base64 encoded public key.
+    #[wasm_bindgen(getter, js_name = "publicKeyBase64")]
+    pub fn public_key(&self) -> JsString {
+        self.inner.to_base64().into()
+    }
+
+    /// The passphrase info, if the key was derived from one.
+    #[wasm_bindgen(getter, js_name = "passphraseInfo")]
+    pub fn passphrase_info(&self) -> Option<PassphraseInfo> {
+        self.passphrase_info.clone()
+    }
+
+    /// Get the full name of the backup algorithm this backup key supports.
+    #[wasm_bindgen(getter, js_name = "algorithm")]
+    pub fn backup_algorithm(&self) -> JsString {
+        self.inner.backup_algorithm().into()
+    }
+
+    /// Signatures that have signed our backup key.
+    /// map of userId to map of deviceOrKeyId to signature
+    #[wasm_bindgen(getter, js_name = "signatures")]
+    pub fn signatures(&self) -> JsValue {
+        let signatures: HashMap<String, HashMap<String, String>> = self
+            .inner
+            .signatures()
+            .into_iter()
+            .map(|(k, v)| (k.to_string(), v.into_iter().map(|(k, v)| (k.to_string(), v)).collect()))
+            .collect();
+
+        serde_wasm_bindgen::to_value(&signatures).unwrap()
```

Comment:
Left an unwrap here.

Reply:
I'm far from convinced that we need this method anyway. What is the application supposed to do with a list of signatures once it has one? It should be up to the rust side to validate the signatures.

Will remove this for now.